### ISSUE TITLE
Reporting Functionality (Weave.jl and Literate.jl)

### ISSUE THREAD

**Issue by cpfiffer** [TuringLang/Turing.jl#517]
This is continued from pull #512. I thought it might be wise to pull this particular conversation into a separate issue in the event that I'm missing something fairly obvious.

There's a problem I have with using both [Weave](https://github.com/mpastell/Weave.jl/) and [Literate](https://github.com/fredrikekre/Literate.jl). `@model` and/or `~` do not play nicely with other Julia code that tries to run it. I have tried to execute them in both Literate and Weave environments, and they error out once I try to invoke a sampler. I've even tried making `@example` blocks in Documenter, which executes code blocks -- same issue. 

Here's an example of the error I get from trying to Weave the `gdemo` model:

````julia
using Turing
using Turing: @~, @model
using MCMCChain, Plots, Distributions

# Define a simple Normal model with unknown mean and variance.
@model gdemo(x) = begin
  s ~ InverseGamma(2,3)
  m ~ Normal(0,sqrt(s))
  x[1] ~ Normal(m, sqrt(s))
  x[2] ~ Normal(m, sqrt(s))
  return s, m
end

#  Run sampler, collect results
c1 = sample(gdemo([1.5, 2]), SMC(1000))
````

<pre class="julia-error">
ERROR: LoadError: UndefVarError: @~ not defined
in expression starting at &lt;macrocall&gt;:-1
</pre>

````julia

# Summarise results (currently requires the master branch from MCMCChain)
describe(c1)
````

<pre class="julia-error">
ERROR: UndefVarError: c1 not defined
</pre>

````julia

# Plot and save results
p = plot(c1)
````

<pre class="julia-error">
ERROR: UndefVarError: c1 not defined
</pre>

Is there something obvious I'm missing here as to why this would fail? If it's just me, or there's only small modifications on my part necessary, I would be delighted. 

It might be my setup, so I'm wondering if we can rule that out quickly. Would someone else try to run the file `weave-it.jl` after installing Weave and copy the contents of the `.md` file here?

[weavetest.zip](https://github.com/TuringLang/Turing.jl/files/2369116/weavetest.zip)

**Comment by mohamed82008** [comment]
Literate seems to work fine. You just need to define a file like this https://github.com/mohamed82008/Turing.jl/blob/literate/docs/example.jl or https://github.com/fredrikekre/Literate.jl/blob/master/examples/README.jl and then call
```julia
using Literate
Literate.markdown("./example.jl", "."; documenter=false)
```
in the directory of the file to generate the md file like in https://github.com/mohamed82008/Turing.jl/blob/literate/docs/example.md. Not sure if you tried that or not. I haven't used Weave before.

**Comment by cpfiffer** [comment]
Literate works by itself because it only copies the code over and formats it to markdown. If I set `documenter = true` and try to execute those blocks via Documenter, that's when the issues occur. 

I can convert the _inputs_ to markdown pretty easily, but I can't actually get the code to execute from within any of the literate programming tools the Julia has.

**Comment by mohamed82008** [comment]
Relevant issue: https://github.com/fredrikekre/Literate.jl/issues/9. If you post your attempt there, you might get better help.

**Comment by mohamed82008** [comment]
With master Documenter, I am getting the same error as you.

**Comment by mohamed82008** [comment]
Most likely some scoping or parsing bug in Documenter.

**Comment by cpfiffer** [comment]
I'll take this over to the Weave people.

**Comment by mpastell** [comment]
Both Weave and Documenter (https://github.com/JuliaDocs/Documenter.jl/blob/c58e7e0fc0d62dc4ce631487002930813e4c51b3/src/DocTests.jl#L186) use a sandbox module to evaluate code blocks, which seems to be the problem here. I posted a solution for Weave in https://github.com/mpastell/Weave.jl/issues/145 .

**Comment by cpfiffer** [comment]
Closed [Weave #145](https://github.com/mpastell/Weave.jl/issues/145). Thank you for your help!
